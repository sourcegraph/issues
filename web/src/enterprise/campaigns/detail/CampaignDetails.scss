.campaign-details {
    &__property-label {
        width: 100px;
    }
    &__property-row {
        align-items: baseline;
    }
}
