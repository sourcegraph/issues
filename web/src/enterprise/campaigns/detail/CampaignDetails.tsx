import { LoadingSpinner } from '@sourcegraph/react-loading-spinner'
import AlertCircleIcon from 'mdi-react/AlertCircleIcon'
import React, { useState, useEffect, useRef, useMemo, useCallback } from 'react'
import * as GQL from '../../../../../shared/src/graphql/schema'
import { HeroPage } from '../../../components/HeroPage'
import { PageTitle } from '../../../components/PageTitle'
import { UserAvatar } from '../../../user/UserAvatar'
import { Timestamp } from '../../../components/time/Timestamp'
import { noop, isEqual } from 'lodash'
import { fetchCampaignById, deleteCampaign, closeCampaign } from './backend'
import { useError } from '../../../../../shared/src/util/useObservable'
import { asError } from '../../../../../shared/src/util/errors'
import * as H from 'history'
import { CampaignBurndownChart } from './BurndownChart'
import { Subject, of, merge, Observable } from 'rxjs'
import { renderMarkdown } from '../../../../../shared/src/util/markdown'
import { ErrorAlert } from '../../../components/alerts'
import { Markdown } from '../../../../../shared/src/components/Markdown'
import { switchMap, distinctUntilChanged } from 'rxjs/operators'
import { ThemeProps } from '../../../../../shared/src/theme'
import { CampaignStatus } from './CampaignStatus'
import { CampaignActionsBar } from './CampaignActionsBar'
import { CampaignChangesets } from './changesets/CampaignChangesets'
import { CampaignDiffStat } from './CampaignDiffStat'
import { pluralize } from '../../../../../shared/src/util/strings'
import { ExtensionsControllerProps } from '../../../../../shared/src/extensions/controller'
import { PlatformContextProps } from '../../../../../shared/src/platform/context'
import { TelemetryProps } from '../../../../../shared/src/telemetry/telemetryService'
import { repeatUntil } from '../../../../../shared/src/util/rxjs/repeatUntil'

export type CampaignUIMode = 'viewing' | 'editing' | 'saving' | 'deleting' | 'closing'

interface Campaign
    extends Pick<
        GQL.ICampaign,
        | '__typename'
        | 'id'
        | 'name'
        | 'description'
        | 'author'
        | 'changesetCountsOverTime'
        | 'createdAt'
        | 'updatedAt'
        | 'closedAt'
        | 'viewerCanAdminister'
        | 'branch'
    > {
    changesets: Pick<GQL.ICampaign['changesets'], 'totalCount'>
    status: Pick<GQL.ICampaign['status'], 'completedCount' | 'pendingCount' | 'errors' | 'state'>
    diffStat: Pick<GQL.ICampaign['diffStat'], 'added' | 'deleted' | 'changed'>
}

interface Props extends ThemeProps, ExtensionsControllerProps, PlatformContextProps, TelemetryProps {
    /**
     * The campaign ID.
     * If not given, will display a creation form.
     */
    campaignID?: GQL.ID
    authenticatedUser: Pick<GQL.IUser, 'id' | 'username' | 'avatarURL'>
    history: H.History
    location: H.Location

    /** For testing only. */
    _fetchCampaignById?: typeof fetchCampaignById | ((campaign: GQL.ID) => Observable<Campaign | null>)
}

/**
 * The area for a single campaign.
 */
export const CampaignDetails: React.FunctionComponent<Props> = ({
    campaignID,
    history,
    location,
    authenticatedUser,
    isLightTheme,
    extensionsController,
    platformContext,
    telemetryService,
    _fetchCampaignById = fetchCampaignById,
}) => {
    // For errors during fetching
    const triggerError = useError()

    /** Retrigger campaign fetching */
    const campaignUpdates = useMemo(() => new Subject<void>(), [])
    /** Retrigger changeset fetching */
    const changesetUpdates = useMemo(() => new Subject<void>(), [])

    const [campaign, setCampaign] = useState<Campaign | null>()
    useEffect(() => {
        if (!campaignID) {
            return
        }
        // on the very first fetch, a reload of the changesets is not required
        let isFirstCampaignFetch = true

        // Fetch campaign if ID was given
        const subscription = merge(of(undefined), campaignUpdates)
            .pipe(
                switchMap(() =>
                    _fetchCampaignById(campaignID).pipe(
                        // repeat fetching the campaign as long as the state is still processing
                        repeatUntil(campaign => campaign?.status?.state !== GQL.BackgroundProcessState.PROCESSING, {
                            delay: 2000,
                        })
                    )
                ),
                distinctUntilChanged((a, b) => isEqual(a, b))
            )
            .subscribe({
                next: fetchedCampaign => {
                    setCampaign(fetchedCampaign)
                    if (!isFirstCampaignFetch) {
                        changesetUpdates.next()
                    }
                    isFirstCampaignFetch = false
                },
                error: triggerError,
            })
        return () => subscription.unsubscribe()
    }, [campaignID, triggerError, changesetUpdates, campaignUpdates, _fetchCampaignById])

    const [mode, setMode] = useState<CampaignUIMode>(campaignID ? 'viewing' : 'editing')

    // To report errors from saving or deleting
    const [alertError, setAlertError] = useState<Error>()

    // To unblock the history after leaving edit mode
    const unblockHistoryReference = useRef<H.UnregisterCallback>(noop)
    useEffect(() => {
        if (!campaignID) {
            unblockHistoryReference.current()
            unblockHistoryReference.current = history.block('Do you want to discard this campaign?')
        }
        // Note: the current() method gets dynamically reassigned,
        // therefor we can't return it directly.
        return () => unblockHistoryReference.current()
    }, [campaignID, history])

    const onClose = useCallback(
        async (closeChangesets: boolean): Promise<void> => {
            if (!confirm('Are you sure you want to close the campaign?')) {
                return
            }
            setMode('closing')
            try {
                await closeCampaign(campaign!.id, closeChangesets)
                campaignUpdates.next()
            } catch (error) {
                setAlertError(asError(error))
            } finally {
                setMode('viewing')
            }
        },
        [campaign, campaignUpdates]
    )

    const onDelete = useCallback(async (): Promise<void> => {
        if (!confirm('Are you sure you want to delete the campaign?')) {
            return
        }
        setMode('deleting')
        try {
            await deleteCampaign(campaign!.id)
            history.push('/campaigns')
        } catch (error) {
            setAlertError(asError(error))
            setMode('viewing')
        }
    }, [campaign, history])

    // Is loading
    if (campaignID && campaign === undefined) {
        return (
            <div className="text-center">
                <LoadingSpinner className="icon-inline mx-auto my-4" />
            </div>
        )
    }
    // Campaign was not found
    if (campaign === null) {
        return <HeroPage icon={AlertCircleIcon} title="Campaign not found" />
    }

    const author = campaign ? campaign.author : authenticatedUser

    const totalChangesetCount = campaign?.changesets.totalCount ?? 0

    const campaignFormID = 'campaign-form'

    return (
        <>
            <PageTitle title={campaign?.name} />
            <CampaignActionsBar
                mode={mode}
                campaign={campaign}
                onClose={onClose}
                onDelete={onDelete}
                formID={campaignFormID}
            />
            {alertError && <ErrorAlert error={alertError} history={history} />}
            {campaign && !['saving', 'editing'].includes(mode) && (
                <CampaignStatus campaign={campaign} history={history} />
            )}
            {campaign && !['saving', 'editing'].includes(mode) && (
                <>
                    <div className="card mt-2">
                        <div className="card-header">
                            <strong>
                                <UserAvatar user={author} className="icon-inline" /> {author.username}
                            </strong>{' '}
                            started <Timestamp date={campaign.createdAt} />
                        </div>
                        <div className="card-body">
                            <Markdown
                                dangerousInnerHTML={renderMarkdown(campaign.description || '_No description_')}
                                history={history}
                            />
                        </div>
                    </div>
                    {totalChangesetCount > 0 && (
                        <>
                            <h3 className="mt-4 mb-2">Progress</h3>
                            <CampaignBurndownChart
                                changesetCountsOverTime={campaign.changesetCountsOverTime}
                                history={history}
                            />
                        </>
                    )}
                </>
            )}

            {totalChangesetCount && (
                <>
                    <h3 className="mt-4 d-flex align-items-end mb-0">
                        {(totalChangesetCount > 0 || !!campaign) && (
                            <>
                                {totalChangesetCount} {pluralize('Changeset', totalChangesetCount)}
                            </>
                        )}{' '}
                        {campaign && <CampaignDiffStat campaign={campaign} className="ml-2 mb-0" />}
                    </h3>
                    {totalChangesetCount > 0 && (
                        <CampaignChangesets
                            campaign={campaign!}
                            changesetUpdates={changesetUpdates}
                            campaignUpdates={campaignUpdates}
                            history={history}
                            location={location}
                            isLightTheme={isLightTheme}
                            extensionsController={extensionsController}
                            platformContext={platformContext}
                            telemetryService={telemetryService}
                        />
                    )}
                </>
            )}
        </>
    )
}
