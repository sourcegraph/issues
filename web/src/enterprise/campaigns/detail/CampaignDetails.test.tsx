import React from 'react'
import { CampaignDetails } from './CampaignDetails'
import * as H from 'history'
import { createRenderer } from 'react-test-renderer/shallow'

jest.mock('../../../settings/MonacoSettingsEditor', () => ({ MonacoSettingsEditor: 'MonacoSettingsEditor' }))

describe('CampaignDetails', () => {
    test('creation form', () => {
        const history = H.createMemoryHistory()
        expect(
            createRenderer().render(
                <CampaignDetails
                    campaignID={undefined}
                    history={history}
                    location={history.location}
                    authenticatedUser={{ username: 'alice', avatarURL: null }}
                    isLightTheme={true}
                />
            )
        ).toMatchSnapshot()
    })
})
