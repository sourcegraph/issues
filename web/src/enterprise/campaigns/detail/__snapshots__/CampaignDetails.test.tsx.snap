// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignDetails creation form for empty manual campaign 1`] = `
<Fragment>
  <PageTitle />
  <CampaignActionsBar
    formID="campaign-form"
    mode="editing"
    onClose={[Function]}
    onDelete={[Function]}
  />
  0
</Fragment>
`;

exports[`CampaignDetails creation form given existing patch set 1`] = `
<CampaignDetails
  _fetchPatchSetById={[Function]}
  authenticatedUser={
    Object {
      "avatarURL": null,
      "id": "a",
      "username": "alice",
    }
  }
  history="[History]"
  isLightTheme={true}
  location="[Location path=/?patchSet=p]"
  telemetryService={
    Object {
      "log": [Function],
      "logViewEvent": [Function],
    }
  }
>
  <PageTitle />
  <CampaignActionsBar
    formID="campaign-form"
    mode="editing"
    onClose={[Function]}
    onDelete={[Function]}
  >
    <div
      className="d-flex mb-2 position-relative"
    >
      <h2
        className="m-0"
      >
        <CampaignsIcon
          className="icon-inline campaign-actions-bar__campaign-icon text-muted mr-2"
        />
        <span>
          <AnchorLink
            to="/campaigns"
          >
            <a
              href="/campaigns"
            >
              Campaigns
            </a>
          </AnchorLink>
           
          <span
            className="badge badge-info"
          >
            Beta
          </span>
        </span>
        <span
          className="text-muted d-inline-block mx-2"
        >
          /
        </span>
        <span>
          New campaign
        </span>
      </h2>
      <span
        className="flex-grow-1 d-flex justify-content-end align-items-center"
      />
    </div>
  </CampaignActionsBar>
  0
</CampaignDetails>
`;

exports[`CampaignDetails viewerCanAdminister: false viewing existing 1`] = `
<CampaignDetails
  _fetchCampaignById={[Function]}
  authenticatedUser={
    Object {
      "avatarURL": null,
      "id": "a",
      "username": "alice",
    }
  }
  campaignID="c"
  history="[History]"
  isLightTheme={true}
  location="[Location path=/]"
  telemetryService={
    Object {
      "log": [Function],
      "logViewEvent": [Function],
    }
  }
>
  <PageTitle
    title="n"
  />
  <CampaignActionsBar
    campaign={
      Object {
        "__typename": "Campaign",
        "author": Object {
          "username": "alice",
        },
        "branch": "awesome-branch",
        "changesetCountsOverTime": Array [],
        "changesets": Object {
          "totalCount": 2,
        },
        "closedAt": null,
        "createdAt": "2020-01-01",
        "description": "d",
        "diffStat": Object {
          "__typename": "IDiffStat",
          "added": 5,
          "changed": 3,
          "deleted": 2,
        },
        "hasUnpublishedPatches": false,
        "id": "c",
        "name": "n",
        "patches": Object {
          "totalCount": 2,
        },
        "status": Object {
          "completedCount": 3,
          "errors": Array [
            "a",
          ],
          "pendingCount": 3,
          "state": "PROCESSING",
        },
        "updatedAt": "2020-01-01",
        "viewerCanAdminister": false,
      }
    }
    formID="campaign-form"
    mode="viewing"
    onClose={[Function]}
    onDelete={[Function]}
  >
    <div
      className="d-flex mb-2 position-relative"
    >
      <h2
        className="m-0"
      >
        <span
          className="badge badge-success mr-2"
        >
          <CampaignsIcon
            className="icon-inline campaign-actions-bar__campaign-icon"
          />
           Open
        </span>
        <span>
          <AnchorLink
            to="/campaigns"
          >
            <a
              href="/campaigns"
            >
              Campaigns
            </a>
          </AnchorLink>
           
          <span
            className="badge badge-info"
          >
            Beta
          </span>
        </span>
        <span
          className="text-muted d-inline-block mx-2"
        >
          /
        </span>
        <span>
          n
        </span>
      </h2>
      <span
        className="flex-grow-1 d-flex justify-content-end align-items-center"
      />
    </div>
  </CampaignActionsBar>
  <CampaignStatus
    campaign={
      Object {
        "__typename": "Campaign",
        "author": Object {
          "username": "alice",
        },
        "branch": "awesome-branch",
        "changesetCountsOverTime": Array [],
        "changesets": Object {
          "totalCount": 2,
        },
        "closedAt": null,
        "createdAt": "2020-01-01",
        "description": "d",
        "diffStat": Object {
          "__typename": "IDiffStat",
          "added": 5,
          "changed": 3,
          "deleted": 2,
        },
        "hasUnpublishedPatches": false,
        "id": "c",
        "name": "n",
        "patches": Object {
          "totalCount": 2,
        },
        "status": Object {
          "completedCount": 3,
          "errors": Array [
            "a",
          ],
          "pendingCount": 3,
          "state": "PROCESSING",
        },
        "updatedAt": "2020-01-01",
        "viewerCanAdminister": false,
      }
    }
    history="[History]"
  >
    CampaignStatus(state=PROCESSING)
  </CampaignStatus>
  <div
    className="card mt-2"
  >
    <div
      className="card-header"
    >
      <strong>
        <UserAvatar
          className="icon-inline"
          user={
            Object {
              "username": "alice",
            }
          }
        />
         
        alice
      </strong>
       
      started 
      <Timestamp
        date="2020-01-01"
      >
        <span
          className="timestamp"
          data-tooltip="2020-01-01"
        >
          in almost 14 years
        </span>
      </Timestamp>
    </div>
    <div
      className="card-body"
    >
      <Markdown
        dangerousInnerHTML="<p>d</p>
"
        history="[History]"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>d</p>
",
            }
          }
          onClick={[Function]}
        />
      </Markdown>
    </div>
  </div>
  <h3
    className="mt-4 mb-2"
  >
    Progress
  </h3>
  <CampaignBurndownChart
    changesetCountsOverTime={Array []}
    history="[History]"
  >
    <p>
      <em>
        Burndown chart will be shown when there is more than 1 day of data.
      </em>
    </p>
  </CampaignBurndownChart>
  <h3
    className="mt-4 d-flex align-items-end mb-0"
  >
    2
     
    Changesets
     
    <CampaignDiffStat
      campaign={
        Object {
          "__typename": "Campaign",
          "author": Object {
            "username": "alice",
          },
          "branch": "awesome-branch",
          "changesetCountsOverTime": Array [],
          "changesets": Object {
            "totalCount": 2,
          },
          "closedAt": null,
          "createdAt": "2020-01-01",
          "description": "d",
          "diffStat": Object {
            "__typename": "IDiffStat",
            "added": 5,
            "changed": 3,
            "deleted": 2,
          },
          "hasUnpublishedPatches": false,
          "id": "c",
          "name": "n",
          "patches": Object {
            "totalCount": 2,
          },
          "status": Object {
            "completedCount": 3,
            "errors": Array [
              "a",
            ],
            "pendingCount": 3,
            "state": "PROCESSING",
          },
          "updatedAt": "2020-01-01",
          "viewerCanAdminister": false,
        }
      }
      className="ml-2 mb-0"
    >
      <div
        className="ml-2 mb-0"
      >
        <DiffStat
          added={5}
          changed={3}
          deleted={2}
          expandedCounts={true}
        >
          <div
            className="diff-stat "
            data-tooltip="5 additions, 3 changes, 2 deletions"
          >
            <span
              className="diff-stat__total font-weight-bold"
            >
              <span
                className="diff-stat__text-added mr-1"
              >
                +
                5
              </span>
              <span
                className="diff-stat__text-changed mr-1"
              >
                •
                3
              </span>
              <span
                className="diff-stat__text-deleted mr-1"
              >
                −
                2
              </span>
            </span>
            <div
              className="diff-stat__square diff-stat__added"
              key="0"
            />
            <div
              className="diff-stat__square diff-stat__added"
              key="1"
            />
            <div
              className="diff-stat__square diff-stat__added"
              key="2"
            />
            <div
              className="diff-stat__square diff-stat__changed"
              key="3"
            />
            <div
              className="diff-stat__square diff-stat__deleted"
              key="4"
            />
          </div>
        </DiffStat>
      </div>
    </CampaignDiffStat>
  </h3>
  <CampaignChangesets
    campaign={
      Object {
        "__typename": "Campaign",
        "author": Object {
          "username": "alice",
        },
        "branch": "awesome-branch",
        "changesetCountsOverTime": Array [],
        "changesets": Object {
          "totalCount": 2,
        },
        "closedAt": null,
        "createdAt": "2020-01-01",
        "description": "d",
        "diffStat": Object {
          "__typename": "IDiffStat",
          "added": 5,
          "changed": 3,
          "deleted": 2,
        },
        "hasUnpublishedPatches": false,
        "id": "c",
        "name": "n",
        "patches": Object {
          "totalCount": 2,
        },
        "status": Object {
          "completedCount": 3,
          "errors": Array [
            "a",
          ],
          "pendingCount": 3,
          "state": "PROCESSING",
        },
        "updatedAt": "2020-01-01",
        "viewerCanAdminister": false,
      }
    }
    campaignUpdates="[Subject]"
    changesetUpdates="[Subject]"
    history="[History]"
    isLightTheme={true}
    location="[Location path=/]"
    telemetryService={
      Object {
        "log": [Function],
        "logViewEvent": [Function],
      }
    }
  />
</CampaignDetails>
`;

exports[`CampaignDetails viewerCanAdminister: true viewing existing 1`] = `
<CampaignDetails
  _fetchCampaignById={[Function]}
  authenticatedUser={
    Object {
      "avatarURL": null,
      "id": "a",
      "username": "alice",
    }
  }
  campaignID="c"
  history="[History]"
  isLightTheme={true}
  location="[Location path=/]"
  telemetryService={
    Object {
      "log": [Function],
      "logViewEvent": [Function],
    }
  }
>
  <PageTitle
    title="n"
  />
  <CampaignActionsBar
    campaign={
      Object {
        "__typename": "Campaign",
        "author": Object {
          "username": "alice",
        },
        "branch": "awesome-branch",
        "changesetCountsOverTime": Array [],
        "changesets": Object {
          "totalCount": 2,
        },
        "closedAt": null,
        "createdAt": "2020-01-01",
        "description": "d",
        "diffStat": Object {
          "__typename": "IDiffStat",
          "added": 5,
          "changed": 3,
          "deleted": 2,
        },
        "hasUnpublishedPatches": false,
        "id": "c",
        "name": "n",
        "patches": Object {
          "totalCount": 2,
        },
        "status": Object {
          "completedCount": 3,
          "errors": Array [
            "a",
          ],
          "pendingCount": 3,
          "state": "PROCESSING",
        },
        "updatedAt": "2020-01-01",
        "viewerCanAdminister": true,
      }
    }
    formID="campaign-form"
    mode="viewing"
    onClose={[Function]}
    onDelete={[Function]}
  >
    <div
      className="d-flex mb-2 position-relative"
    >
      <h2
        className="m-0"
      >
        <span
          className="badge badge-success mr-2"
        >
          <CampaignsIcon
            className="icon-inline campaign-actions-bar__campaign-icon"
          />
           Open
        </span>
        <span>
          <AnchorLink
            to="/campaigns"
          >
            <a
              href="/campaigns"
            >
              Campaigns
            </a>
          </AnchorLink>
           
          <span
            className="badge badge-info"
          >
            Beta
          </span>
        </span>
        <span
          className="text-muted d-inline-block mx-2"
        >
          /
        </span>
        <span>
          n
        </span>
      </h2>
      <span
        className="flex-grow-1 d-flex justify-content-end align-items-center"
      >
        <CloseDeleteCampaignPrompt
          buttonClassName="btn-secondary mr-1"
          buttonText="Close"
          disabled={true}
          disabledTooltip="Cannot close while campaign is being created"
          message={
            <p>
              Close campaign 
              <strong>
                n
              </strong>
              ?
            </p>
          }
          onButtonClick={[Function]}
        >
          <details
            className="campaign-prompt__details"
          >
            <summary>
              <span
                className="btn dropdown-toggle btn-secondary mr-1 disabled"
                data-tooltip="Cannot close while campaign is being created"
                onClick={[Function]}
              >
                Close
              </span>
            </summary>
            <div
              className="position-absolute campaign-prompt__details-menu"
            >
              <div
                className="card mt-1"
              >
                <div
                  className="card-body"
                >
                  <p>
                    Close campaign 
                    <strong>
                      n
                    </strong>
                    ?
                  </p>
                  <div
                    className="form-group"
                  >
                    <label>
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="checkbox"
                      />
                       
                      Close open changesets on code hosts
                    </label>
                  </div>
                  <button
                    className="btn mr-1 btn-secondary mr-1"
                    disabled={true}
                    onClick={[Function]}
                    type="button"
                  >
                    Close
                  </button>
                </div>
              </div>
            </div>
          </details>
        </CloseDeleteCampaignPrompt>
        <CloseDeleteCampaignPrompt
          buttonClassName="btn-danger"
          buttonText="Delete"
          disabled={true}
          disabledTooltip="Cannot delete while campaign is being created"
          message={
            <p>
              Delete campaign 
              <strong>
                n
              </strong>
              ?
            </p>
          }
          onButtonClick={[Function]}
        >
          <details
            className="campaign-prompt__details"
          >
            <summary>
              <span
                className="btn dropdown-toggle btn-danger disabled"
                data-tooltip="Cannot delete while campaign is being created"
                onClick={[Function]}
              >
                Delete
              </span>
            </summary>
            <div
              className="position-absolute campaign-prompt__details-menu"
            >
              <div
                className="card mt-1"
              >
                <div
                  className="card-body"
                >
                  <p>
                    Delete campaign 
                    <strong>
                      n
                    </strong>
                    ?
                  </p>
                  <div
                    className="form-group"
                  >
                    <label>
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="checkbox"
                      />
                       
                      Close open changesets on code hosts
                    </label>
                  </div>
                  <button
                    className="btn mr-1 btn-danger"
                    disabled={true}
                    onClick={[Function]}
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </details>
        </CloseDeleteCampaignPrompt>
      </span>
    </div>
  </CampaignActionsBar>
  <CampaignStatus
    campaign={
      Object {
        "__typename": "Campaign",
        "author": Object {
          "username": "alice",
        },
        "branch": "awesome-branch",
        "changesetCountsOverTime": Array [],
        "changesets": Object {
          "totalCount": 2,
        },
        "closedAt": null,
        "createdAt": "2020-01-01",
        "description": "d",
        "diffStat": Object {
          "__typename": "IDiffStat",
          "added": 5,
          "changed": 3,
          "deleted": 2,
        },
        "hasUnpublishedPatches": false,
        "id": "c",
        "name": "n",
        "patches": Object {
          "totalCount": 2,
        },
        "status": Object {
          "completedCount": 3,
          "errors": Array [
            "a",
          ],
          "pendingCount": 3,
          "state": "PROCESSING",
        },
        "updatedAt": "2020-01-01",
        "viewerCanAdminister": true,
      }
    }
    history="[History]"
  >
    CampaignStatus(state=PROCESSING)
  </CampaignStatus>
  <div
    className="card mt-2"
  >
    <div
      className="card-header"
    >
      <strong>
        <UserAvatar
          className="icon-inline"
          user={
            Object {
              "username": "alice",
            }
          }
        />
         
        alice
      </strong>
       
      started 
      <Timestamp
        date="2020-01-01"
      >
        <span
          className="timestamp"
          data-tooltip="2020-01-01"
        >
          in almost 14 years
        </span>
      </Timestamp>
    </div>
    <div
      className="card-body"
    >
      <Markdown
        dangerousInnerHTML="<p>d</p>
"
        history="[History]"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>d</p>
",
            }
          }
          onClick={[Function]}
        />
      </Markdown>
    </div>
  </div>
  <h3
    className="mt-4 mb-2"
  >
    Progress
  </h3>
  <CampaignBurndownChart
    changesetCountsOverTime={Array []}
    history="[History]"
  >
    <p>
      <em>
        Burndown chart will be shown when there is more than 1 day of data.
      </em>
    </p>
  </CampaignBurndownChart>
  <h3
    className="mt-4 d-flex align-items-end mb-0"
  >
    2
     
    Changesets
     
    <CampaignDiffStat
      campaign={
        Object {
          "__typename": "Campaign",
          "author": Object {
            "username": "alice",
          },
          "branch": "awesome-branch",
          "changesetCountsOverTime": Array [],
          "changesets": Object {
            "totalCount": 2,
          },
          "closedAt": null,
          "createdAt": "2020-01-01",
          "description": "d",
          "diffStat": Object {
            "__typename": "IDiffStat",
            "added": 5,
            "changed": 3,
            "deleted": 2,
          },
          "hasUnpublishedPatches": false,
          "id": "c",
          "name": "n",
          "patches": Object {
            "totalCount": 2,
          },
          "status": Object {
            "completedCount": 3,
            "errors": Array [
              "a",
            ],
            "pendingCount": 3,
            "state": "PROCESSING",
          },
          "updatedAt": "2020-01-01",
          "viewerCanAdminister": true,
        }
      }
      className="ml-2 mb-0"
    >
      <div
        className="ml-2 mb-0"
      >
        <DiffStat
          added={5}
          changed={3}
          deleted={2}
          expandedCounts={true}
        >
          <div
            className="diff-stat "
            data-tooltip="5 additions, 3 changes, 2 deletions"
          >
            <span
              className="diff-stat__total font-weight-bold"
            >
              <span
                className="diff-stat__text-added mr-1"
              >
                +
                5
              </span>
              <span
                className="diff-stat__text-changed mr-1"
              >
                •
                3
              </span>
              <span
                className="diff-stat__text-deleted mr-1"
              >
                −
                2
              </span>
            </span>
            <div
              className="diff-stat__square diff-stat__added"
              key="0"
            />
            <div
              className="diff-stat__square diff-stat__added"
              key="1"
            />
            <div
              className="diff-stat__square diff-stat__added"
              key="2"
            />
            <div
              className="diff-stat__square diff-stat__changed"
              key="3"
            />
            <div
              className="diff-stat__square diff-stat__deleted"
              key="4"
            />
          </div>
        </DiffStat>
      </div>
    </CampaignDiffStat>
  </h3>
  <CampaignChangesets
    campaign={
      Object {
        "__typename": "Campaign",
        "author": Object {
          "username": "alice",
        },
        "branch": "awesome-branch",
        "changesetCountsOverTime": Array [],
        "changesets": Object {
          "totalCount": 2,
        },
        "closedAt": null,
        "createdAt": "2020-01-01",
        "description": "d",
        "diffStat": Object {
          "__typename": "IDiffStat",
          "added": 5,
          "changed": 3,
          "deleted": 2,
        },
        "hasUnpublishedPatches": false,
        "id": "c",
        "name": "n",
        "patches": Object {
          "totalCount": 2,
        },
        "status": Object {
          "completedCount": 3,
          "errors": Array [
            "a",
          ],
          "pendingCount": 3,
          "state": "PROCESSING",
        },
        "updatedAt": "2020-01-01",
        "viewerCanAdminister": true,
      }
    }
    campaignUpdates="[Subject]"
    changesetUpdates="[Subject]"
    history="[History]"
    isLightTheme={true}
    location="[Location path=/]"
    telemetryService={
      Object {
        "log": [Function],
        "logViewEvent": [Function],
      }
    }
  />
</CampaignDetails>
`;
